Unbox types properly from arrays

workingjubilee opened the pull request:
Introducing `Datum<'src>` allows using it with HRTB fns in schema generation to persist lifetimes all the way through `#[pg_extern]`'s wrapper-fn infrastructure without compromising the integrity of lifetime annotations. This is immediately implemented to prevent many types from exceeding the lifetime of the Array they are unboxed from, unless they are already "self-owned" and thus 'static.

Unfortunately, this does not fix all instances using `pg_getarg`, and those simply are left unsolved for now in some cases. I reorganized the UI tests to make it easier to notice these problems and when they get solved.

workingjubilee committed:
Add `Datum<'dat>`

workingjubilee committed:
Add UnboxDatum

workingjubilee committed:
Rampage over Array

workingjubilee committed:
Make arbitrary PostgresEnums unboxable

workingjubilee committed:
Disable unsoundness tests

They no longer pass borrowck.

workingjubilee committed:
Fix lifetime on json test

workingjubilee committed:
Sorta-mostly fix arrays kinda

workingjubilee committed:
Lift impl of FunctionMetadata to F: Fn(A) -> R

workingjubilee committed:
Allow fn pointers to need hrtb

workingjubilee committed:
preserve some code

workingjubilee committed:
Stop rewriting types to have zero spaces

workingjubilee committed:
Anonymize lts in used types

workingjubilee committed:
Use the named lifetime for composite_type!

workingjubilee committed:
Autofixes

workingjubilee committed:
rustfmt

workingjubilee committed:
Remove staticize_lifetimes

workingjubilee committed:
Remove broken staticize_lifetimes code

workingjubilee committed:
Remove inline staticization in used_type.rs

workingjubilee committed:
Simplify FunctionMetadata impls for fn() -> R

workingjubilee committed:
Allow using FnMut

workingjubilee committed:
Commit the TypeId crime

workingjubilee committed:
well apparently that works

workingjubilee committed:
Erase unused code and fmt

workingjubilee committed:
Unify split_for_impl since no staticization

workingjubilee committed:
Comment on reasons

workingjubilee committed:
Make arrays work with lifetimes

Add a lifetime bound to array iterators, so serialize works.

workingjubilee committed:
Adding bounds to heap_tuple unboxing

workingjubilee committed:
Add unboxing for VariadicArray

workingjubilee committed:
Add notes on work-to-do

workingjubilee committed:
Docs and inlining hints for UnboxDatum

workingjubilee committed:
Document more

workingjubilee committed:
Add array leakage test

workingjubilee committed:
Revert "Make arrays work with lifetimes"

This reverts commit 6106a144ddc0a899f2601634b2e84bb7c6a96c51.

workingjubilee committed:
Make another bad example into a ui test

workingjubilee committed:
Comment out the broken tests

workingjubilee committed:
Try to fix array iteration

workingjubilee committed:
Futz with lifetimes

workingjubilee committed:
Remove spare file

workingjubilee committed:
Revert "Remove spare file"

This reverts commit e4118f81326c28d0456cb8bd4bcf4202ad6c9193.

workingjubilee committed:
Revert "Futz with lifetimes"

This reverts commit aa2274d1aebca356eb0a597065031c02703808a2.

workingjubilee committed:
Revert "Try to fix array iteration"

This reverts commit b2cf0300ba4fb377eb0ff15def629fa8922b32dc.

workingjubilee committed:
Try to fix htup lifetimes

workingjubilee committed:
Add unboxing for PostgresTypes

workingjubilee committed:
Prevent leaking lifetimes from HeapTuple

workingjubilee committed:
Fix postgresenum impl

workingjubilee committed:
Mark currently-unusable tests

workingjubilee committed:
Get over the line with some failing tests

workingjubilee committed:
Infest arrays with pedantically correct lifetimes

workingjubilee committed:
Fix last doctest

workingjubilee committed:
Erase redundant lifetime

workingjubilee committed:
fmt

workingjubilee committed:
More lifetime pedantry

workingjubilee committed:
More cleanup and strictly static

workingjubilee committed:
Fix test more properly

workingjubilee committed:
Fix doc comment

workingjubilee committed:
Minor cleanup

workingjubilee committed:
Fix most LT testing in roundtrip tests

workingjubilee commented:
I can split this down the line of [a0184f4](https://github.com/pgcentralfoundation/pgrx/pull/1476/commits/a0184f4edeafbb52d7ac37da7738e78a512c5d1d) and [e8ead32](https://github.com/pgcentralfoundation/pgrx/pull/1476/commits/e8ead32f62b80805dfbbb37044c20b2d9a4fecac) if you want, and I can add more notes explaining the lifetimes, names, etc. but this seems like it would benefit from a cursory glance to inform that step, at least.

workingjubilee committed:
Add some TODOs

eeeebbbbrrrr commented on pgrx-sql-entity-graph/src/lifetimes.rs:
```diff
@@ -7,76 +7,6 @@
-    }
-}
-
-pub fn staticize_lifetimes(value: &mut syn::Type) {
```
goodby, `staticize_lifetimes`... you will not be missed!

eeeebbbbrrrr commented on pgrx-examples/bad_ideas/src/lib.rs:
```diff
@@ -55,7 +55,7 @@ fn warning(s: &str) -> bool {
 fn exec<'a>(
     command: &'a str,
-    args: default!(Vec<Option<&'a str>>, "ARRAY[]::text[]"),
+    args: default!(Vec<Option<String>>, "ARRAY[]::text[]"),
```
Just a question... why this change?

Could it not be `Vec<Option<&'a str>>`?

This was the only example you touched, so either that means this example is the only one that accepts a `&str` and that pattern is no longer valid, or you changed this for some other reason?

workingjubilee replied:
Because this changes FromDatum in the impls in pgrx/src/datum/array.rs to be in terms of `T: UnboxDatum` with certain bounds, this PR currently breaks `FromDatum` specifically for `Vec<Option<T>>` where there is any doubt as to whether `T` implements `UnboxDatum`. Essentially, the trait resolver can't figure it out at that depth. This can be solved by involving additional traits which work differently and either allow `Vec<Option<T>>` to not depend directly on `Array<'_, T>` for its unboxing or to move `pg_getarg` (or its replacement) off `T: FromDatum`.

eeeebbbbrrrr reviewed (approved):
Jubilee, this looks really clean.  It's clear you put a lot of thought into this to basically only have a few lifetime signature changes throughout the code.

Speaking of which, I'm surprised PgHeapTuple didn't require more effort than this.  Cool.

workingjubilee commented:
> Speaking of which, I'm surprised PgHeapTuple didn't require more effort than this. Cool.

GitHub won't open it implicitly because it is a "large diff", but unfortunately, pgrx-tests/src/tests/heap_tuple.rs betrays how many cases won't work now, and so this isn't really finished work. I had a PR similar to this a couple months ago, which did (more) damage, but I needed all the work until now to understand whether it would be viable to even fix, and to fix a number of less-complex cases that hinged on that groundwork. I need to add in a bunch more changes to make all of the cases work that need to work properly. Some of those hinge on SPI, which started busted and remains busted.

...Ultimately I am unconcerned about making `PgHeapTuple<'a, WA>` work really _well_, because I now believe it's the wrong abstraction in the long run, so as a warning, I have accepted most of those examples getting broken on a semi-permanent basis.

workingjubilee committed:
Add todo tests as ui tests

workingjubilee committed:
Mv compile-fail tests to a dir

workingjubilee committed:
Add serde test to todo tests

workingjubilee committed:
Add roundtrip tests to todo

workingjubilee commented:
I added the currently-broken tests to the set of ui tests, and reorganized tests slightly, so that it will be easier for us to notice when those problems are fixed.

workingjubilee committed:
Fixup todo tests

workingjubilee commented on pgrx-tests/tests/ui/heap-tuples-dont-leak.rs:
```diff
@@ -0,0 +1,29 @@
+    let dog: PgHeapTuple<AllocatedByRust> = dog;
+
+    dog.get_by_name("name").unwrap().unwrap()
+}
```
```suggestion
}

```

workingjubilee commented on pgrx-tests/tests/compile-fail/heap-tuples-dont-leak.rs:
```diff
@@ -0,0 +1,29 @@
+    let dog: PgHeapTuple<AllocatedByRust> = dog;
+
+    dog.get_by_name("name").unwrap().unwrap()
+}
```
```suggestion
}

```

workingjubilee commented on pgrx-tests/tests/todo/array-serialize-json.rs:
```diff
@@ -0,0 +1,32 @@
+        assert_eq!(json.0, json! {{"values": ["one", null, "two", "three"]}});
+        Ok(())
+    }
+}
```
```suggestion
}

```

workingjubilee committed:
Add newlines because GH's EOF symbol annoys me

workingjubilee committed:
More test fixup

workingjubilee committed:
More fixes to test todos

workingjubilee committed:
Rename 'dat to 'src

Lifetimes should be named for their upper bound, and Datum<'dat>
is bound on the source of the Datum.

workingjubilee committed:
fmt
